python/paddle/fluid/transpiler/memory_optimization_transpiler.py

```diff
@@ -398,6 +475,19 @@ def memory_optimize(input_program, skip_opt_set=None, print_log=False, level=0):
         raise ValueError("only support opt_level 0 or 1.")
     global PRINT_LOG
     PRINT_LOG = print_log
+    if skip_grads:
+        grad_set = set()
+        OP_ROLE_VAR = core.op_proto_and_checker_maker.kOpRoleVarAttrName()
+        for op in input_program.global_block().ops:
+            if _is_opt_role_op(op):
+                if op.attr(OP_ROLE_VAR):
+                    grad_name = op.attr(OP_ROLE_VAR)[1]
```

Comment:
Why the optimize vars grad can not be reused?  
Take SGD for example,  w@Grad, b@Grad these things are temporary variables, reuse logic should be applied to these ones, right?

Reply:
Not all grad vars are skipped, only grad vars that have the corresponding parameter. 

We need to skip these vars when running distributed training, or else the parameter server will generate optimize blocks reusing vars like when we have L2 regularizers, pserver will generate one block for each parameter optimization, and run them in parallel, then these reused vars will be accessed in multiple threads.

Since these vars cannot be reused in psever, then we just skip all parameters' grad vars, which will be send to parameter server. And for pserver, we just use the original program to get parallel optimization blocks.

Reply:
Have discussed with @typhoonzero offline.